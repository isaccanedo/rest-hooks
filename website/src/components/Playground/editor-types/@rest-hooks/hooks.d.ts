import { EndpointInterface, EndpointParam } from '@rest-hooks/endpoint';

/**
 * Keeps value updated after delay time
 *
 * @see https://resthooks.io/docs/api/useDebounce
 * @param value Any immutable value
 * @param delay Time in miliseconds to wait til updating the value
 * @param updatable Whether to update at all
 * @example
 ```
 const debouncedFilter = useDebounced(filter, 200);
 const list = useResource(ListShape, { filter });
 ```
 */
declare function useDebounce<T>(value: T, delay: number, updatable?: boolean): T;

/**
 * Builds an Endpoint that cancels fetch everytime params change
 *
 * @see https://resthooks.io/docs/api/useCancelling
 * @example
 ```
 useResource(useCancelling(MyEndpoint, { id }), { id })
 ```
 */
declare function useCancelling<E extends EndpointInterface & {
    extend: (o: {
        signal?: AbortSignal | undefined;
    }) => any;
}>(endpoint: E, params: EndpointParam<E> | null): E;

/**
 * Takes an async function and tracks resolution as a boolean.
 *
 * @see https://resthooks.io/docs/api/useLoading
 * @param func A function returning a promise
 * @param deps Deps list sent to useCallback()
 * @example
 ```
 function Button({ onClick, children, ...props }) {
   const [clickHandler, loading] = useLoading(onClick);
   return (
     <button onClick={clickHandler} {...props}>
       {loading ? 'Loading...' : children}
     </button>
   );
 }
 ```
 */
declare function useLoading<F extends (...args: any) => Promise<any>>(func: F, deps?: readonly any[]): [F, boolean, Error | undefined];

export { useCancelling, useDebounce, useLoading };
