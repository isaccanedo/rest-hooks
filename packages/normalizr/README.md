# normalizr
[![CircleCI](https://circleci.com/gh/coinbase/rest-hooks/tree/master.svg?style=shield)](https://circleci.com/gh/coinbase/rest-hooks)
[![Coverage Status](https://img.shields.io/codecov/c/gh/coinbase/rest-hooks/master.svg?style=flat-square)](https://app.codecov.io/gh/coinbase/rest-hooks?branch=master)
[![npm version](https://img.shields.io/npm/v/@rest-hooks/normalizr.svg?style=flat-square)](https://www.npmjs.com/package/@rest-hooks/normalizr) [![npm downloads](https://img.shields.io/npm/dm/@rest-hooks/normalizr.svg?style=flat-square)](https://www.npmjs.com/package/@rest-hooks/normalizr)

## Install

Install from the NPM repository using yarn or npm:

```shell
yarn add @rest-hooks/normalizr
```

```shell
npm install @rest-hooks/normalizr
```

## Motivation

Many APIs, public or not, return JSON data that has deeply nested objects. Using data in this kind of structure is often very difficult for JavaScript applications, especially those using [Flux](http://facebook.github.io/flux/) or [Redux](http://redux.js.org/).

## Solution

Normalizr is a small, but powerful utility for taking JSON with a schema definition and returning nested entities with their IDs, gathered in dictionaries.

## Documentation

* [Introduction](./docs/README.md)
  * [Build Files](./docs/README.md#build-files)
* [Quick Start](https://resthooks.io/docs/api/schema)
* [API](https://resthooks.io/docs/api/Entity)
  * [normalize](./docs/api.md#normalizedata-schema)
  * [denormalize](./docs/api.md#denormalizeinput-schema-entities)
  * [schema](https://resthooks.io/docs/api/schema)

## Examples

* [Normalizing GitHub Issues](/examples/normalizr-github)
* [Relational Data](/examples/normalizr-relationships)
* [Interactive Redux](/examples/normalizr-redux)

## Quick Start

Consider a typical blog post. The API response for a single post might look something like this:

```json
{
  "id": "123",
  "author": {
    "id": "1",
    "name": "Paul"
  },
  "title": "My awesome blog post",
  "comments": [
    {
      "id": "324",
      "commenter": {
        "id": "2",
        "name": "Nicole"
      }
    }
  ]
}
```

We have two nested entity types within our `article`: `users` and `comments`. Using various `schema`, we can normalize all three entity types down:

```js
import { normalize, schema, Entity } from '@rest-hooks/normalizr';

// Define a users schema
class User extends Entity {
  pk() { return this.id }
}

// Define your comments schema
class Comment extends Entity {
  pk() { return this.id }

  static schema = {
    commenter: User,
  }
}

// Define your article
class Article extends Entity {
  pk() { return this.id }

  static schema = {
  author: User,
  comments: [Comment]
  }
}

const normalizedData = normalize(originalData, Article);
```

Now, `normalizedData` will be:

```js
{
  result: "123",
  entities: {
    "Article": {
      "123": {
        id: "123",
        author: "1",
        title: "My awesome blog post",
        comments: [ "324" ]
      }
    },
    "User": {
      "1": { "id": "1", "name": "Paul" },
      "2": { "id": "2", "name": "Nicole" }
    },
    "Comment": {
      "324": { id: "324", "commenter": "2" }
    }
  }
}
```

## Dependencies

None.

## Credits

Normalizr was originally created by [Dan Abramov](http://github.com/gaearon) and inspired by a conversation with [Jing Chen](https://twitter.com/jingc). Since v3, it was completely rewritten and maintained by [Paul Armstrong](https://twitter.com/paularmstrong).
Since v4, it was largely rewritten and maintained by [Nathaniel Tucker](https://twitter.com/npinp).
It has also received much help, enthusiasm, and contributions from [community members](https://github.com/ntucker/normalizr/graphs/contributors).
