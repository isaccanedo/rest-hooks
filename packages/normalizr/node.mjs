export * from './dist/normalizr.js';
