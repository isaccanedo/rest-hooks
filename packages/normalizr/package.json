{
  "name": "@rest-hooks/normalizr",
  "version": "9.2.0",
  "description": "Normalizes and denormalizes JSON according to schema for Redux and Flux applications",
  "homepage": "https://github.com/coinbase/rest-hooks/tree/master/packages/normalizr#readme",
  "bugs": {
    "url": "https://github.com/coinbase/rest-hooks/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:coinbase/rest-hooks.git",
    "directory": "packages/normalizr"
  },
  "keywords": [
    "flux",
    "redux",
    "normalize",
    "denormalize",
    "data",
    "cache",
    "api",
    "react-native",
    "ios",
    "android",
    "web",
    "json"
  ],
  "files": [
    "dist/",
    "lib/",
    "node.mjs",
    "legacy",
    "ts3.4",
    "LICENSE",
    "README.md"
  ],
  "main": "dist/normalizr.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/normalizr.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "sideEffects": false,
  "scripts": {
    "build": "yarn run build:lib && yarn run build:legacy:lib && yarn run build:bundle",
    "build:js:development": "BROWSERSLIST_ENV=legacy NODE_ENV=development rollup -c",
    "build:js:production": "BROWSERSLIST_ENV=legacy NODE_ENV=production rollup -c",
    "build:js:node": "BROWSERSLIST_ENV=node12 NODE_ENV=development rollup -c",
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' babel --root-mode upward src --out-dir lib --source-maps inline --extensions '.ts,.tsx,.js' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' babel --root-mode upward src --out-dir legacy --source-maps inline --extensions '.ts,.tsx,.js' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:bundle": "run-s build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "rimraf lib ts3.4 legacy dist *.tsbuildinfo",
    "build:legacy-types": "yarn run downlevel-dts lib ts3.4",
    "lint": "yarn lint:cmd --fix",
    "lint:ci": "yarn lint:cmd",
    "lint:cmd": "eslint . --ext '.js,.json,.snap' --cache",
    "precommit": "flow check && lint-staged",
    "prepublishOnly": "npm run build"
  },
  "author": "Nathaniel Tucker",
  "contributors": [
    "Paul Armstrong",
    "Dan Abramov"
  ],
  "license": "MIT",
  "devDependencies": {
    "@babel/cli": "7.19.3",
    "@babel/core": "7.20.2",
    "@types/babel__core": "^7",
    "downlevel-dts": "^0.10.0",
    "immutable": "4.1.0",
    "npm-run-all": "^4.1.5",
    "rollup": "2.79.1",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-filesize": "^9.1.2",
    "rollup-plugin-json": "^4.0.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-terser": "^7.0.2"
  },
  "dependencies": {
    "@babel/runtime": "^7.13.0",
    "npm-run-all": "^4.1.5"
  }
}
