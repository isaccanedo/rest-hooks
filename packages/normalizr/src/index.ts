import { denormalize } from './denormalize.js';
import { normalize } from './normalize.js';
import WeakListMap from './WeakListMap.js';
import { isEntity } from './isEntity.js';
export { default as inferResults } from './inferResults.js';
export { DELETED } from './special.js';

export type {
  AbstractInstanceType,
  NormalizeReturnType,
  NormalizedSchema,
  DenormalizeReturnType,
  DenormalizeCache,
} from './types.js';
export * from './endpoint/types.js';
export * from './interface.js';
export * from './Expiry.js';
export * from './normal.js';

export { denormalize, normalize, isEntity, WeakListMap };
