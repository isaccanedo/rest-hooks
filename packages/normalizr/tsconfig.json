{
  "extends": "../../tsconfig-base",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src",
    "noImplicitAny": false,
    "skipLibCheck": false
  },
  "include": ["src"]
}
