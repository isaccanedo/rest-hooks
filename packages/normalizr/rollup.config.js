import babel from 'rollup-plugin-babel';
import commonjs from 'rollup-plugin-commonjs';
import filesize from 'rollup-plugin-filesize';
import resolve from 'rollup-plugin-node-resolve';
import { terser } from 'rollup-plugin-terser';
import dts from 'rollup-plugin-dts';

import { name } from './package.json';

const isProduction = process.env.NODE_ENV === 'production';

const destBase = 'dist/normalizr';
const destExtension = `${isProduction ? '.min' : ''}.js`;
const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node'];

function snakeCase(id) {
  return id.replace(/(-|\/)/g, '_').replace(/@/g, '');
}

const configs = [];

if (process.env.BROWSERSLIST_ENV !== 'node12') {
  if (process.env.NODE_ENV === 'production') {
    configs.push({
      input: 'lib/index.js',
      output: [{ file: `${destBase}.es${destExtension}`, format: 'es' }],
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          runtimeHelpers: true,
          rootMode: 'upward',
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
  } else {
    configs.push({
      input: 'lib/index.js',
      output: [
        {
          file: `${destBase}.umd${destExtension}`,
          format: 'umd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.amd${destExtension}`,
          format: 'amd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.browser${destExtension}`,
          format: 'iife',
          name: snakeCase(name),
        },
      ],
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          runtimeHelpers: true,
          rootMode: 'upward',
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
    configs.push({
      input: './lib/index.d.ts',
      output: [{ file: 'index.d.ts', format: 'es' }],
      plugins: [dts()],
    });
  }
} else {
  // node-friendly commonjs build
  configs.push({
    input: 'lib/index.js',
    output: [{ file: `${destBase}${destExtension}`, format: 'cjs' }],
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        runtimeHelpers: true,
        rootMode: 'upward',
      }),
      resolve({ extensions }),
      commonjs({ extensions }),
      isProduction && terser(),
      filesize(),
    ].filter(Boolean),
  });
}
export default configs;
