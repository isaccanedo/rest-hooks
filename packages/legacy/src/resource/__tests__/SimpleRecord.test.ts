import { denormalize, normalize, DELETED } from '@rest-hooks/normalizr';

import { Entity, SimpleRecord } from '..';

let dateSpy: jest.SpyInstance;
beforeAll(() => {
  dateSpy = jest
    // eslint-disable-next-line no-undef
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

class Article extends SimpleRecord {
  readonly id: string = '';
  readonly title: string = '';
  readonly author: string = '';
  readonly content: string = '';
}
class ArticleEntity extends Entity {
  readonly id: string = '';
  readonly title: string = '';
  readonly author: string = '';
  readonly content: string = '';
  pk() {
    return this.id;
  }
}

class Pagination extends SimpleRecord {
  readonly data = ArticleEntity.fromJS();
  readonly nextPage: string = '';
  readonly lastPage: string = '';

  get title() {
    return this.data.title;
  }

  static schema = {
    data: ArticleEntity,
  };
}

class WithOptional extends SimpleRecord {
  readonly article: ArticleEntity | null = null;
  readonly requiredArticle = ArticleEntity.fromJS();
  readonly nextPage = '';
  readonly createdAt: Date | null = null;

  static schema = {
    article: ArticleEntity,
    requiredArticle: ArticleEntity,
    createdAt: Date,
  };
}

describe('SimpleRecord', () => {
  it('should init', () => {
    const resource = Article.fromJS({
      id: '5',
      title: 'happy',
      author: '5',
    });
    expect(resource.title).toBe('happy');
    expect(resource.author).toBe('5');
  });

  it('should not allow init with wrong types', () => {
    Article.fromJS({
      // @ts-expect-error
      id: 5,
      title: 'happy',
      // @ts-expect-error
      author: 5,
    });
  });

  describe('normalize', () => {
    it('should normalize into plain object', () => {
      const schema = Article;
      const normalized = normalize(
        {
          id: '5',
          title: 'happy',
          author: '5',
        },
        schema,
      );
      expect(normalized.result).toEqual({
        id: '5',
        title: 'happy',
        author: '5',
      });
    });

    it('should work on nested', () => {
      const schema = Pagination;
      const normalized = normalize(
        {
          data: {
            id: '5',
            title: 'happy',
            author: '5',
          },
          lastPage: 'abc',
          nextPage: 'slksfd',
        },
        schema,
      );
      expect(normalized).toMatchSnapshot();
    });

    it('should deserialize Date', () => {
      const normalized = normalize(
        {
          requiredArticle: { id: '5' },
          nextPage: 'blob',
          createdAt: '2020-06-07T02:00:15.000Z',
        },
        WithOptional,
      );
      expect(normalized.result.createdAt.getTime()).toBe(
        normalized.result.createdAt.getTime(),
      );
      expect(normalized).toMatchSnapshot();
    });

    it('should use default when Date not provided', () => {
      const normalized = normalize(
        {
          requiredArticle: { id: '5' },
          nextPage: 'blob',
        },
        WithOptional,
      );
      expect(normalized.result.createdAt).toBeUndefined();
      expect(normalized).toMatchSnapshot();
    });
  });

  describe('denormalize', () => {
    it('should denormalize with defaults', () => {
      const schema = Article;
      const denormalized = denormalize(
        {
          id: '5',
          title: 'happy',
          author: '5',
        },
        schema,
        {},
      );
      const article = denormalized[0];
      expect(article).toBeDefined();
      expect(article).toBeInstanceOf(Article);
      expect(article).toEqual({
        id: '5',
        title: 'happy',
        author: '5',
        content: '',
      });
      // typing of members that it has
      article?.content;
      article?.author;
      // @ts-expect-error - fails when it doesn't
      article?.bob;
    });

    it('should work with nested and default correctly', () => {
      const schema = Pagination;
      const denormalized = denormalize(
        {
          data: '5',
          nextPage: 'blob',
        },
        schema,
        {
          [ArticleEntity.key]: {
            ['5']: {
              id: '5',
              title: 'happy',
              author: '5',
            },
          },
        },
      );
      expect(denormalized[1]).toBe(true);
      const pagination = denormalized[0];
      expect(pagination).toBeDefined();
      expect(pagination).toBeInstanceOf(Pagination);
      expect(pagination).toEqual({
        data: {
          id: '5',
          title: 'happy',
          author: '5',
          content: '',
        },
        nextPage: 'blob',
        lastPage: '',
      });
      // check getters
      expect(pagination.title).toBe('happy');
      // typing of members that it has
      pagination.data?.content;
      pagination.data?.title;
      // @ts-expect-error - fails when it doesn't have member
      pagination?.bob;
    });
  });

  it('should have undefined member when entity is not found', () => {
    const schema = Pagination;
    const denormalized = denormalize(
      {
        data: '5',
        nextPage: 'blob',
      },
      schema,
      {},
    );
    expect(denormalized[1]).toBe(false);
    const pagination = denormalized[0];
    expect(pagination).toBeDefined();
    expect(pagination).toBeInstanceOf(Pagination);
    expect(pagination).toEqual({
      data: undefined,
      nextPage: 'blob',
      lastPage: '',
    });
  });

  describe('optional entities', () => {
    it('should be marked as found even when optional is not there', () => {
      const denormalized = denormalize(
        {
          requiredArticle: '5',
          nextPage: 'blob',
          createdAt: new Date('2020-06-07T02:00:15+0000'),
        },
        WithOptional,
        {
          [ArticleEntity.key]: {
            5: ArticleEntity.fromJS({ id: '5' }),
          },
        },
      );
      expect(denormalized[1]).toBe(true);
      const response = denormalized[0];
      expect(response).toBeDefined();
      expect(response).toBeInstanceOf(WithOptional);
      expect(response).toEqual({
        article: null,
        requiredArticle: ArticleEntity.fromJS({ id: '5' }),
        nextPage: 'blob',
        createdAt: new Date('2020-06-07T02:00:15+0000'),
      });
      // @ts-expect-error
      response.createdAt.toISOString();
      expect(response.createdAt?.toISOString()).toBe(
        '2020-06-07T02:00:15.000Z',
      );
    });

    it('should be marked as not found when required entity is missing', () => {
      const denormalized = denormalize(
        {
          article: '5',
          nextPage: 'blob',
        },
        WithOptional,
        {
          [ArticleEntity.key]: {
            5: ArticleEntity.fromJS({ id: '5' }),
          },
        },
      );
      expect(denormalized[1]).toBe(false);
      expect(denormalized[2]).toBe(false);
      const response = denormalized[0];
      expect(response).toBeDefined();
      expect(response).toBeInstanceOf(WithOptional);
      expect(response).toEqual({
        article: ArticleEntity.fromJS({ id: '5' }),
        requiredArticle: ArticleEntity.fromJS(),
        nextPage: 'blob',
        createdAt: null,
      });
      expect(response.createdAt).toBeNull();
    });

    it('should be marked as deleted when required entity is deleted symbol', () => {
      const denormalized = denormalize(
        {
          article: '5',
          requiredArticle: '6',
          nextPage: 'blob',
        },
        WithOptional,
        {
          [ArticleEntity.key]: {
            5: ArticleEntity.fromJS({ id: '5' }),
            6: DELETED,
          },
        },
      );
      expect(denormalized[1]).toBe(true);
      expect(denormalized[2]).toBe(true);
      const response = denormalized[0];
      expect(response).toBeDefined();
      expect(response).toBeInstanceOf(WithOptional);
      expect(response).toEqual({
        article: ArticleEntity.fromJS({ id: '5' }),
        requiredArticle: undefined,
        nextPage: 'blob',
        createdAt: null,
      });
      expect(response.createdAt).toBeNull();
    });
  });
});
