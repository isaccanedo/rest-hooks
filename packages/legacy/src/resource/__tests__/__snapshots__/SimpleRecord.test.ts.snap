// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleRecord normalize should deserialize Date 1`] = `
{
  "entities": {
    "ArticleEntity": {
      "5": ArticleEntity {
        "author": "",
        "content": "",
        "id": "5",
        "title": "",
      },
    },
  },
  "entityMeta": {
    "ArticleEntity": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "createdAt": 2020-06-07T02:00:15.000Z,
    "nextPage": "blob",
    "requiredArticle": "5",
  },
}
`;

exports[`SimpleRecord normalize should use default when Date not provided 1`] = `
{
  "entities": {
    "ArticleEntity": {
      "5": ArticleEntity {
        "author": "",
        "content": "",
        "id": "5",
        "title": "",
      },
    },
  },
  "entityMeta": {
    "ArticleEntity": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "nextPage": "blob",
    "requiredArticle": "5",
  },
}
`;

exports[`SimpleRecord normalize should work on nested 1`] = `
{
  "entities": {
    "ArticleEntity": {
      "5": ArticleEntity {
        "author": "5",
        "content": "",
        "id": "5",
        "title": "happy",
      },
    },
  },
  "entityMeta": {
    "ArticleEntity": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "data": "5",
    "lastPage": "abc",
    "nextPage": "slksfd",
  },
}
`;
